vllm/v1/worker/gpu_model_runner.py

```diff
@@ -1017,20 +1019,56 @@ def execute_model(
         scheduler_output: "SchedulerOutput",
         intermediate_tensors: Optional[IntermediateTensors] = None,
     ) -> Union[ModelRunnerOutput, torch.Tensor]:
-        # Update KVConnector with the KVConnector metadata forward().
-        if has_kv_transfer_group():
-            get_kv_transfer_group().bind_connector_metadata(
-                scheduler_output.kv_connector_metadata)
+
+        def maybe_setup_kv_connector():
+            # Update KVConnector with the KVConnector metadata forward().
+            if has_kv_transfer_group():
+                kv_connector = get_kv_transfer_group()
+                assert isinstance(kv_connector, KVConnectorBase_V1)
+                assert scheduler_output.kv_connector_metadata is not None
+                kv_connector.bind_connector_metadata(
+                    scheduler_output.kv_connector_metadata)
+
+                # Background KV cache transfers happen here.
+                # These transfers are designed to be async and the requests
+                # involved may be disjoint from the running requests.
+                # Do this here to save a collective_rpc.
+                kv_connector.start_load_kv(get_forward_context())
+
+        def maybe_wait_for_save():
+            if has_kv_transfer_group():
+                kv_connector = get_kv_transfer_group()
+                kv_connector.wait_for_save()
+
+        def maybe_get_finished() -> tuple[set[str], set[str]]:
+            if has_kv_transfer_group():
+                kv_connector = get_kv_transfer_group()
+                return kv_connector.get_finished()
+            else:
+                return set(), set()
 
         self._update_states(scheduler_output)
         if not scheduler_output.total_num_scheduled_tokens:
+            # KV send/recv even if no work to do.
+            with set_forward_context(None, self.vllm_config):
+                maybe_setup_kv_connector()
+                maybe_wait_for_save()
+                finished_sending, finished_recving = maybe_get_finished()
+
             # Return empty ModelRunnerOutput if there's no work to do.
-            return EMPTY_MODEL_RUNNER_OUTPUT
+            output = EMPTY_MODEL_RUNNER_OUTPUT
+
+            if len(finished_sending) > 0 or len(finished_recving) > 0:
+                output = copy.deepcopy(EMPTY_MODEL_RUNNER_OUTPUT)
```

Comment:
no need for deep? (@mgoin is the expert on this)
```suggestion
                output = copy.copy(EMPTY_MODEL_RUNNER_OUTPUT)
```

Reply:
You're right deep is not needed. It think we should do a deepcopy though to be safe since if someone appends to `output.finished_sending` then this would require a deepcopy